### PR TITLE
Migrate legacy thumbnails (fixes #235)

### LINKED ISSUE #235: File Migration: Reuse existing resized images
## Overview

 * As a developer migrating from 3.x to 4.x, I want to minimise the time spent on migration tasks, in order to de-risk my go-live plan
 * As a sysadmin migrating from 3.x to 4.x, I want to keep the system requirements at the same level in order to avoid provisioning more costly resources (e.g. more disk space)
 * As a CMS author using an upgraded 4.x CMS, I want to manage assets without being slowed down by system tasks

## Acceptance Criteria

 * The file migration task reuses existing resized images on the filesystem (and moves/renames them as required)
 * The file migration task generates new resized images where required (CMS thumbnails), but allows developers to opt out of this
 * The file migration task does not increase the disk space required for assets

## Notes

 * Assumes that missing CMS thumbnails are lazy generated on demand
 * We'll need to update the [file migration reference](https://docs.silverstripe.org/en/4/developer_guides/files/file_migration/#migration-of-thumbnails) to thumbnail behaviour

## Pull Requests

 * [x] https://github.com/silverstripe/silverstripe-framework/pull/8924
 * [x] https://github.com/silverstripe/silverstripe-assets/pull/242

### PR DESCRIPTION
**Pull request by chillu**
Merge after https://github.com/silverstripe/silverstripe-framework/pull/8924

Relies on @max https://github.com/silverstripe/silverstripe-assets/pull/223 to be merged first. The newest commits by me are the actual PR here

### CODE REVIEW

**Comment by maxime-rainville** [comment]
Travis failure is caused by the main file migration task which hasn't been converted to use the strategies yet.

File: src/LegacyThumbnailMigrationHelper.php [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        // Get normalised store relative path
+        $folderPath = preg_replace('#^/#', '', $folder->getFilename());
+        $resampledFolderPath = $folderPath . '_resampled';
+
+        // Legacy thumbnails couldn't have been stored in a protected filesystem
+        /** @var \League\Flysystem\Filesystem $filesystem */
+        $filesystem = $store->getPublicFilesystem();
+
+        /** @var FileResolutionStrategy $strategy */
+        $strategy = $store->getPublicResolutionStrategy();
```
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: Done. Changed the dependency from the `AssetStore` interface to the `FlysystemAssetStore` concrete class. Technically we could create another interface for those additional methods, but I think there's limited value in that extra API maintenance overhead just to get a bit more purity in a helper task.

**Comment by tractorcow** [comment]
That's a lot of code!

Is there something in particular you want me to review? I can't review all of this but I can look at specifics if you need input.

File: src/LegacyThumbnailMigrationHelper.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+        // Recurse through folder
+        foreach ($filesystem->listContents($resampledFolderPath, true) as $fileInfo) {
+            if ($fileInfo['type'] !== 'file') {
+                continue;
+            }
+
+            $oldResampledPath = $fileInfo['path'];
+
+            // Get "variant folders" inside _resampled folder.
+            $variantFolders = preg_replace('#.*_resampled/(.*)#', '$1', $fileInfo['dirname']);
```
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: Dunno, I'd need to explode the resulting variant on underscores then, instead of just dealing with the directories. So double conversion and more moving parts in order to remove hardcoding `_resampled` in one place - feels low value.

File: src/LegacyThumbnailMigrationHelper.php [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+
+            // Replicate new variant format.
+            // We're always placing these files in the public filesystem, *without* a content hash path.
+            // This means you need to run the optional migration task introduced in 4.4,
+            // which moves public files out of content hash folders.
+            // Image->manipulate() is in charge of creating the full file name (incl. variant),
+            // and assumes the manipulation is run on an existing original file, so we can't use it here.
+            // Any AssetStore-level filesystem operations (like move()) suffer from the same limitation,
+            // so we need to drop down to path based filesystem renames.
+            $newResampledPath = implode('', [
```
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: No, the point was that it shouldn't matter if the new file exists. It might have been deleted in the meantime. Due to the shortcodes issues (https://github.com/silverstripe/silverstripe-assets/issues/167), existing thumbnails have lost their connection to the original file. So its entirely possible that thumbnails are still relied on in e.g. HTML content, but no longer available through the original file. Technically those are now orphaned files with no way to clean them up through SS APIs, but that's better than breaking images on website content.

**Comment by chillu** [comment]
@tractorcow Awesome, thanks for the offer! Maybe just think about gnarly edge cases that we might've missed? Stuff like "chained format thumbnail file names", "multiple files in same _resampled folder". 

> I'm thinking maybe add a test case for two image files in the same folder with variants sharing the same _resampled sub folder.

Hah, funny I actually had that case in there (`nested-sibling.jpg`), but the fixture handling got too complicated so I trimmed it back down. Will have another look

File: src/LegacyThumbnailMigrationHelper.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+
+            $filesystem->rename($oldResampledPath, $newResampledPath);
+
+            $this->logger->info(sprintf('Moved legacy thumbnail %s to %s', $oldResampledPath, $newResampledPath));
+
+            $moved[$oldResampledPath] = $newResampledPath;
+        }
+
+        // Remove folder and any subfolders
+        $filesystem->deleteDir($resampledFolderPath);
```
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: The `_resampled` folder is specifically not empty: It contains empty subfolders (one for each transformation hash). Looking through the logic, there's nothing in there which skips over files - they're all handled, either through `$filesystem->delete()` or the `$filesystem->rename()`. If it fails to do that for some reason, presumably an exception will stop processing. 

In the unlikely case that you've added your own files into `_resampled` (e.g. custom file generation following a different format from thumbnails), then the logic will still rename those. E.g. `_resampled/my-custom-file-format123.jpg` would be renamed to `my-custom-file-format123__.jpg`. That would probably break your custom logic, but if you used a system folder to place your own files, all bets are off, right?

**Comment by chillu** [comment]
@tractorcow Oh I just realised you might not have picked up on the fact that most commits in this PR are pulled in from a separate PR: https://github.com/silverstripe/silverstripe-assets/pull/223. That other PR is actually more important to get your input on (it refactors a lot of the asset abstraction behaviour), the thumbnail task here is just another cleanup thing that's less important.

File: src/LegacyThumbnailMigrationHelper.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/LegacyThumbnailMigrationHelper.php]: `LegacyFileIDHelper` already swaps the `/` for underscores.

File: src/LegacyThumbnailMigrationHelper.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (maxime-rainville) [reply src/LegacyThumbnailMigrationHelper.php]: I don't see that. Have you pushed your commit?

File: src/LegacyThumbnailMigrationHelper.php [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+ * Example file format:
+ * Before (3.x): `assets/my-folder/_resampled/PadWzYwLDYwLCJGRkZGRkYiLDBd/FillWzYwLDMwXQ/my-image.jpg`
+ * After (4.x): `assets/my-folder/0ec62bd1c4/my-image__PadWzYwLDYwLCJGRkZGRkYiLDBd_CropHeightWzMwXQ.jpg`
+ *
+ * Limitations:
+ * - Does not migrate legacy thumbnails where the original file or folder
+ *   has been renamed since an earlier 4.x migration run
+ * - Does not filter out unused CMS thumbnails (they're using a new size now)
+ */
+class LegacyThumbnailMigrationHelper
```
Comment (maxime-rainville) [inline src/LegacyThumbnailMigrationHelper.php]: We're starting to have quite a few migration test in assets. I'm starting to think maybe we should move them all to their own sub-namespace and alias the existing ones.

Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: Done!

**Comment by chillu** [comment]
> Maybe add an assertion that the variants don't get regenerated when the developer asks for them later on. That's more of a nice to have.

Yeah, I don't see how that would happen - the locations of generated thumbnails should be asserted as part of the existing image manipulation tests, has nothing to do with the migration task, right? Anyway, I've added `ImageTest->testGenerateImageInSameFolderAsOriginal()`

Ready for re-review :)

File: src/LegacyThumbnailMigrationHelper.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: Yeah that's my point: I need the variants individually, not in `variant1/variant2` or `variant1_variant2` format. So I need to mess with strings anyway, even if I'm using `LegacyFileIDHelper`

File: src/LegacyThumbnailMigrationHelper.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (chillu) [reply src/LegacyThumbnailMigrationHelper.php]: Yeah sorry, pushed now. Github needs a "review answer draft" heh

**Comment by maxime-rainville** [comment]
I've merge 1.4 back into this.

File: src/Migration/LegacyThumbnailMigrationHelper.php [hunk 6]
```diff
@@ -0,0 +1,198 @@
+<?php
+
+namespace SilverStripe\Assets\Migration;
```
Comment (maxime-rainville) [inline src/Migration/LegacyThumbnailMigrationHelper.php]: Needs to be updated to SilverStripe\Assets\Dev\Tasks

File: tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php [hunk 7]
```diff
@@ -100,7 +126,10 @@ public function testMigratesThumbnailsForProtectedFiles()
                 new ParsedFileID($image->Filename, $image->Hash),
                 $store->getProtectedFilesystem()
             );
-        $this->assertCount(1, $protectedVariants);
+
+        // findVariants() returns *both* the original file and the variant
+        // See FileIDHelperResolutionStrategyTest->testFindVariant()
+        $this->assertCount(2, $protectedVariants);
```
Comment (maxime-rainville) [inline tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php]: `findVariants` returns a generator. Technically, those are not countable. PHPUnit seems to get around that by iterating over the generator until it reaches the end or goes over the limit. That does meant that when the test fails, you always get a "Failed asserting that actual size 0 matches expected size 2." even if the actual size is something other than zero.

I'm not going any where with this ... just pointing it out. I was half way going down the rabbit hole when I realised that it wasn't a problem.

Comment (maxime-rainville) [reply tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php]: Actually, it is broken.

The moved variant is on the public store. You are looking on the protected store, so you are not getting the variant at all.

You are getting the main file twice:
* Once because it matches the NaturalFileIDHelper
* Once because it matches the LegacyFileIDHelper

File: tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php [hunk 8]
```diff
@@ -68,21 +67,48 @@ public function tearDown()
...
+     */
     public function testMigratesThumbnailsForProtectedFiles()
     {
         /** @var TestAssetStore $store */
         $store = singleton(AssetStore::class); // will use TestAssetStore
 
+        // Remove the legacy helper, otherwise it'll move the _resampled files as well,
+        // which "pollutes" our test case.
+        /** @var FileIDHelperResolutionStrategy $strategy */
+        $publicStrategy = $store->getPublicResolutionStrategy();
```
[the remark this answers is not in the document]
Comment (chillu) [reply tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php]: Hm, I wanted to specifically test with the defaults, *apart* from the `LegacyFileIDHelper`. So if those defaults ever change, that change should be tested through this unit test. So I think my way is more accurate :)

[the remark this answers is not in the document]
Comment (chillu) [reply tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php]: There's a separate angle to this discussion: How much of what we call "unit tests" (ha ha ha) should rely on global configuration which can change on a project-by-project basis. That's currently enshrined through `$backend->setPublicResolutionStrategy(Injector::inst()->get(FileResolutionStrategy::class . '.public'));` in `TestAssetStore` though, so I'm not keen to change it at this late stage since it'll affect all kinds of tests.

**Comment by chillu** [comment]
I've decided to remove a test around thumbnails on protected files, since it's hitting diminishing returns: Probably spent a day on getting all the stars aligned in the various abstractions of abstractions in there. The test currently fails, which correctly points out that this feature isn't implemented. I've documented it as a known limitation instead. It's an edge case.

> Does not move legacy thumbnails to the protected store if the original file has been unpublished or protected since an earlier 4.x migration run

```php
/**
     * Tests cases where files have already been migrated to 4.3 or older,
     * and some images have since been unpublished or otherwise marked protected.
     * The previously orphaned thumbnails wouldn't have been moved to the protected store in this case.
     * The migration task should find those orphans, and reunite them with the original image.
     * And they lived happily ever after.
     */
    public function testMigratesThumbnailsForProtectedFiles()
    {
        /** @var TestAssetStore $store */
        $store = singleton(AssetStore::class); // will use TestAssetStore

        // Remove the legacy helper, otherwise it'll move the _resampled files as well,
        // which "pollutes" our test case.
        /** @var FileIDHelperResolutionStrategy $strategy */
        $publicStrategy = $store->getPublicResolutionStrategy();
        $helpers = $publicStrategy->getResolutionFileIDHelpers();
        $origHelpers = [];
        foreach ($helpers as $i => $helper) {
            $origHelpers[] = clone $helper;
            if ($helper instanceof LegacyFileIDHelper) {
                unset($helpers[$i]);
            }
        }
        $publicStrategy->setResolutionFileIDHelpers($helpers);
        $store->setPublicResolutionStrategy($publicStrategy);

        /** @var Image $image */
        $image = $this->objFromFixture(Image::class, 'nested');

        $formats = ['ResizedImage' => [60,60]];
        $expectedLegacyPath = $this->createLegacyResampledImageFixture($store, $image, $formats);

        // Protect file *after* creating legacy fixture,
        // but without moving the _resampled "orphan"
        $image->protectFile();
        $image->write();

        // We need to retain the legacy helper for later assertions.
        $publicStrategy->setResolutionFileIDHelpers($origHelpers);
        $store->setPublicResolutionStrategy($publicStrategy);

        $helper = new LegacyThumbnailMigrationHelper();
        $moved = $helper->run($store);

        $this->assertCount(1, $moved);

        $publicVariants = $store->getPublicResolutionStrategy()
            ->findVariants(
                new ParsedFileID($image->Filename, $image->Hash),
                $store->getPublicFilesystem()
            );
        $this->assertCount(0, $publicVariants);

        $protectedVariants = $store->getProtectedResolutionStrategy()
            ->findVariants(
                new ParsedFileID($image->Filename, $image->Hash),
                $store->getProtectedFilesystem()
            );

        // findVariants() returns *both* the original file and the variant
        // See FileIDHelperResolutionStrategyTest->testFindVariant()
        // TODO This is counting the original file twice, doesn't include the actual variant
        // See https://github.com/silverstripe/silverstripe-assets/pull/242/files#r279615136
        $this->assertCount(1, $protectedVariants);
    }
```

File: tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php (hunk 7 above)
Comment (chillu) [reply tests/php/Dev/Tasks/LegacyThumbnailMigrationHelperTest.php]: Aaaah great catch! This is a super gnarly test. I've removed it with fa4b14c013b2b71c6984ac03466f8c1c3c599c5b, and added it as a known limitation. See comment on main thread for more rationale: https://github.com/silverstripe/silverstripe-assets/pull/242#issuecomment-488155213
